// @midscene/shared/dist/es/env/parse-model-config.mjs
import { DEFAULT_MODEL_CONFIG_KEYS, INSIGHT_MODEL_CONFIG_KEYS, PLANNING_MODEL_CONFIG_KEYS } from "./constants.mjs";
import { MIDSCENE_OPENAI_HTTP_PROXY, MIDSCENE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_OPENAI_SOCKS_PROXY, MIDSCENE_USE_DOUBAO_VISION, MIDSCENE_USE_GEMINI, MIDSCENE_USE_QWEN3_VL, MIDSCENE_USE_QWEN_VL, MIDSCENE_USE_VLM_UI_TARS, MODEL_FAMILY_VALUES, OPENAI_API_KEY, OPENAI_BASE_URL, UITarsModelVersion } from "./types.mjs";
import { getDebug } from "../logger.mjs";
import { assert } from "../utils.mjs";
import { maskConfig, parseJson } from "./helper.mjs";
import { initDebugConfig } from "./init-debug.mjs";
const KEYS_MAP = {
    insight: INSIGHT_MODEL_CONFIG_KEYS,
    planning: PLANNING_MODEL_CONFIG_KEYS,
    default: DEFAULT_MODEL_CONFIG_KEYS
};
const modelFamilyToVLConfig = (modelFamily)=>{
    if (!modelFamily) return {
        vlMode: void 0,
        uiTarsVersion: void 0
    };
    if ('vlm-ui-tars' === modelFamily) return {
        vlMode: 'vlm-ui-tars',
        uiTarsVersion: UITarsModelVersion.V1_0
    };
    if ('vlm-ui-tars-doubao' === modelFamily || 'vlm-ui-tars-doubao-1.5' === modelFamily) return {
        vlMode: 'vlm-ui-tars',
        uiTarsVersion: UITarsModelVersion.DOUBAO_1_5_20B
    };
    if (!MODEL_FAMILY_VALUES.includes(modelFamily)) throw new Error(`Invalid MIDSCENE_MODEL_FAMILY value: ${modelFamily}`);
    return {
        vlMode: modelFamily,
        uiTarsVersion: void 0
    };
};
const legacyConfigToModelFamily = (provider)=>{
    const isDoubao = provider[MIDSCENE_USE_DOUBAO_VISION];
    const isQwen = provider[MIDSCENE_USE_QWEN_VL];
    const isQwen3 = provider[MIDSCENE_USE_QWEN3_VL];
    const isUiTars = provider[MIDSCENE_USE_VLM_UI_TARS];
    const isGemini = provider[MIDSCENE_USE_GEMINI];
    const enabledModes = [
        isDoubao && MIDSCENE_USE_DOUBAO_VISION,
        isQwen && MIDSCENE_USE_QWEN_VL,
        isQwen3 && MIDSCENE_USE_QWEN3_VL,
        isUiTars && MIDSCENE_USE_VLM_UI_TARS,
        isGemini && MIDSCENE_USE_GEMINI
    ].filter(Boolean);
    if (enabledModes.length > 1) throw new Error(`Only one vision mode can be enabled at a time. Currently enabled modes: ${enabledModes.join(', ')}. Please disable all but one mode.`);
    if (isQwen3) return 'qwen3-vl';
    if (isQwen) return 'qwen2.5-vl';
    if (isDoubao) return 'doubao-vision';
    if (isGemini) return 'gemini';
    if (isUiTars) if ('1' === isUiTars) return 'vlm-ui-tars';
    else if ('DOUBAO' === isUiTars || 'DOUBAO-1.5' === isUiTars) return 'vlm-ui-tars-doubao-1.5';
    else return 'vlm-ui-tars-doubao';
};
const parseOpenaiSdkConfig = ({ keys, provider, useLegacyLogic = false })=>{
    initDebugConfig();
    const debugLog = getDebug('ai:config');
    debugLog('enter parseOpenaiSdkConfig with keys:', keys);
    const legacyAPIKey = useLegacyLogic ? provider[OPENAI_API_KEY] : void 0;
    const legacyBaseURL = useLegacyLogic ? provider[OPENAI_BASE_URL] : void 0;
    const legacySocksProxy = useLegacyLogic ? provider[MIDSCENE_OPENAI_SOCKS_PROXY] : void 0;
    const legacyHttpProxy = useLegacyLogic ? provider[MIDSCENE_OPENAI_HTTP_PROXY] : void 0;
    const legacyOpenaiExtraConfig = useLegacyLogic ? provider[MIDSCENE_OPENAI_INIT_CONFIG_JSON] : void 0;
    const legacyModelFamily = useLegacyLogic ? legacyConfigToModelFamily(provider) : void 0;
    const modelFamilyRaw = provider[keys.modelFamily] || legacyModelFamily;
    const openaiApiKey = provider[keys.openaiApiKey] || legacyAPIKey;
    const openaiBaseURL = provider[keys.openaiBaseURL] || legacyBaseURL;
    const socksProxy = provider[keys.socksProxy] || legacySocksProxy;
    const httpProxy = provider[keys.httpProxy] || legacyHttpProxy;
    const modelName = provider[keys.modelName];
    const openaiExtraConfigStr = provider[keys.openaiExtraConfig];
    const openaiExtraConfig = parseJson(keys.openaiExtraConfig, openaiExtraConfigStr || legacyOpenaiExtraConfig);
    const temperature = provider[keys.temperature] ? Number(provider[keys.temperature]) : 0;
    const { vlMode, uiTarsVersion } = modelFamilyToVLConfig(modelFamilyRaw);
    const getModelDescription = (vlMode, uiTarsVersion)=>{
        if (vlMode) if (uiTarsVersion) return `UI-TARS=${uiTarsVersion}`;
        else return `${vlMode} mode`;
        return '';
    };
    const modelDescription = getModelDescription(vlMode, uiTarsVersion);
    return {
        socksProxy,
        httpProxy,
        vlModeRaw: vlMode,
        openaiBaseURL,
        openaiApiKey,
        openaiExtraConfig,
        vlMode,
        uiTarsModelVersion: uiTarsVersion,
        modelName: modelName,
        modelDescription,
        intent: '-',
        timeout: provider[keys.timeout] ? Number(provider[keys.timeout]) : void 0,
        temperature
    };
};
const decideModelConfigFromIntentConfig = (intent, configMap)=>{
    const debugLog = getDebug('ai:config');
    debugLog('will decideModelConfig base on agent.modelConfig()', intent, maskConfig(configMap));
    const keysForFn = KEYS_MAP[intent];
    const modelName = configMap[keysForFn.modelName];
    if (!modelName) return void debugLog('no modelName found for intent', intent);
    const finalResult = parseOpenaiSdkConfig({
        keys: keysForFn,
        provider: configMap,
        useLegacyLogic: 'default' === intent
    });
    finalResult.intent = intent;
    debugLog('decideModelConfig result by agent.modelConfig() with intent', intent, maskConfig({
        ...finalResult
    }));
    assert(finalResult.openaiBaseURL, `failed to get base URL of model (intent=${intent}). See https://midscenejs.com/model-strategy`);
    if (!finalResult.modelName) console.warn(`modelName is not set for intent ${intent}, this may cause unexpected behavior. See https://midscenejs.com/model-strategy`);
    return finalResult;
};
export { decideModelConfigFromIntentConfig, legacyConfigToModelFamily, modelFamilyToVLConfig, parseOpenaiSdkConfig };
