// @midscene/core/dist/es/ai-model/conversation-history.mjs
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
var _computedKey;
_computedKey = Symbol.iterator;
let _computedKey1 = _computedKey;
class ConversationHistory {
    resetPendingFeedbackMessageIfExists() {
        if (this.pendingFeedbackMessage) this.pendingFeedbackMessage = '';
    }
    append(message) {
        this.messages.push(message);
    }
    seed(messages) {
        this.reset();
        messages.forEach((message)=>{
            this.append(message);
        });
    }
    reset() {
        this.messages.length = 0;
    }
    snapshot(maxImages) {
        if (void 0 === maxImages) return [
            ...this.messages
        ];
        const clonedMessages = structuredClone(this.messages);
        let imageCount = 0;
        for(let i = clonedMessages.length - 1; i >= 0; i--){
            const message = clonedMessages[i];
            const content = message.content;
            if (Array.isArray(content)) for(let j = 0; j < content.length; j++){
                const item = content[j];
                if ('image_url' === item.type) {
                    imageCount++;
                    if (imageCount > maxImages) content[j] = {
                        type: 'text',
                        text: '(image ignored due to size optimization)'
                    };
                }
            }
        }
        return clonedMessages;
    }
    get length() {
        return this.messages.length;
    }
    [_computedKey1]() {
        return this.messages[Symbol.iterator]();
    }
    toJSON() {
        return this.snapshot();
    }
    constructor(options){
        _define_property(this, "messages", []);
        _define_property(this, "pendingFeedbackMessage", void 0);
        if (options?.initialMessages?.length) this.seed(options.initialMessages);
        this.pendingFeedbackMessage = '';
    }
}
export { ConversationHistory };
