// @midscene/web/dist/es/mcp-tools.mjs
import { z } from "@midscene/core";
import { BaseMidsceneTools } from "@midscene/shared/mcp";
import { AgentOverChromeBridge } from "./bridge-mode/index.mjs";
import { StaticPage } from "./static/index.mjs";
class WebMidsceneTools extends BaseMidsceneTools {
    createTemporaryDevice() {
        return new StaticPage({
            screenshotBase64: '',
            size: {
                width: 1920,
                height: 1080
            }
        });
    }
    async ensureAgent(openNewTabWithUrl) {
        if (this.agent && openNewTabWithUrl) {
            try {
                await this.agent?.destroy?.();
            } catch (error) {
                console.debug('Failed to destroy agent during re-init:', error);
            }
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        if (!openNewTabWithUrl) throw new Error('Bridge mode requires a URL. Use web_connect tool to connect to a page first.');
        this.agent = await this.initBridgeModeAgent(openNewTabWithUrl);
        return this.agent;
    }
    async initBridgeModeAgent(url) {
        const agent = new AgentOverChromeBridge({
            closeConflictServer: true
        });
        if (url) await agent.connectNewTabWithUrl(url);
        else await agent.connectCurrentTab();
        return agent;
    }
    preparePlatformTools() {
        return [
            {
                name: 'web_connect',
                description: 'Connect to web page by opening new tab with URL',
                schema: {
                    url: z.string().url().describe('URL to connect to')
                },
                handler: async (args)=>{
                    const { url } = args;
                    const agent = await this.ensureAgent(url);
                    const screenshot = await agent.page?.screenshotBase64();
                    if (!screenshot) return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to: ${url}`
                            }
                        ]
                    };
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to: ${url}`
                            },
                            ...this.buildScreenshotContent(screenshot)
                        ]
                    };
                },
                autoDestroy: false
            }
        ];
    }
}
export { WebMidsceneTools };
