// @midscene/shared/dist/es/extractor/dom-util.mjs
function isFormElement(node) {
    return node instanceof HTMLElement && ('input' === node.tagName.toLowerCase() || 'textarea' === node.tagName.toLowerCase() || 'select' === node.tagName.toLowerCase() || 'option' === node.tagName.toLowerCase());
}
function isButtonElement(node) {
    return node instanceof HTMLElement && 'button' === node.tagName.toLowerCase();
}
function isAElement(node) {
    return node instanceof HTMLElement && 'a' === node.tagName.toLowerCase();
}
function isSvgElement(node) {
    return node instanceof SVGElement;
}
function isImgElement(node) {
    if (!includeBaseElement(node) && node instanceof Element) {
        const computedStyle = window.getComputedStyle(node);
        const backgroundImage = computedStyle.getPropertyValue('background-image');
        if ('none' !== backgroundImage) return true;
    }
    if (isIconfont(node)) return true;
    return node instanceof HTMLElement && 'img' === node.tagName.toLowerCase() || node instanceof SVGElement && 'svg' === node.tagName.toLowerCase();
}
function isIconfont(node) {
    if (node instanceof Element) {
        const computedStyle = window.getComputedStyle(node);
        const fontFamilyValue = computedStyle.fontFamily || '';
        return fontFamilyValue.toLowerCase().indexOf('iconfont') >= 0;
    }
    return false;
}
function isNotContainerElement(node) {
    return isTextElement(node) || isIconfont(node) || isImgElement(node) || isButtonElement(node) || isAElement(node) || isFormElement(node);
}
function isTextElement(node) {
    if (node instanceof Element) {
        if (node?.childNodes?.length === 1 && node?.childNodes[0] instanceof Text) return true;
    }
    return node.nodeName?.toLowerCase?.() === '#text' && !isIconfont(node);
}
function isContainerElement(node) {
    if (!(node instanceof HTMLElement)) return false;
    if (includeBaseElement(node)) return false;
    const computedStyle = window.getComputedStyle(node);
    const backgroundColor = computedStyle.getPropertyValue('background-color');
    if (backgroundColor) return true;
    return false;
}
function includeBaseElement(node) {
    if (!(node instanceof HTMLElement)) return false;
    if (node.innerText) return true;
    const includeList = [
        'svg',
        'button',
        'input',
        'textarea',
        'select',
        'option',
        'img',
        'a'
    ];
    for (const tagName of includeList){
        const element = node.querySelectorAll(tagName);
        if (element.length > 0) return true;
    }
    return false;
}
function generateElementByPosition(position, description) {
    const edgeSize = 8;
    const rect = {
        left: Math.round(Math.max(position.x - edgeSize / 2, 0)),
        top: Math.round(Math.max(position.y - edgeSize / 2, 0)),
        width: edgeSize,
        height: edgeSize
    };
    const element = {
        rect,
        center: [
            Math.round(position.x),
            Math.round(position.y)
        ],
        description: description || ''
    };
    return element;
}
export { generateElementByPosition, isAElement, isButtonElement, isContainerElement, isFormElement, isImgElement, isNotContainerElement, isSvgElement, isTextElement };
